src/cmap/mod.rs

```diff
@@ -57,38 +54,139 @@ impl From<Arc<ConnectionPoolInner>> for ConnectionPool {
     }
 }
 
-/// The internal state of a connection pool.
-#[derive(Derivative)]
-#[derivative(Debug)]
-pub(crate) struct ConnectionPoolInner {
-    /// The address the pool's connections will connect to.
-    address: StreamAddress,
-
+/// A struct used to manage the creation, closing, and storage of connections for a `ConnectionPool`.
+#[derive(Debug)]
+struct ConnectionManager {
     /// The set of available connections in the pool. Because the CMAP spec requires that
     /// connections are checked out in a FIFO manner, connections are pushed/popped from the back
     /// of the Vec.
-    connections: Arc<RwLock<Vec<Connection>>>,
+    checked_in_connections: Vec<Connection>,
+
+    /// The total number of connections managed by the pool, including connections which are
+    /// currently checked out of the pool.
+    total_connection_count: u32,
+
+    /// The ID of the next connection created by the pool.
+    next_connection_id: u32,
+
+    /// The current generation of the pool. The generation is incremented whenever the pool is
+    /// cleared. Connections belonging to a previous generation are considered stale and will be
+    /// closed when checked back in or when popped off of the set of available connections.
+    generation: u32,
 
-    /// The connect timeout passed to each underlying TcpStream when attemtping to connect to the
-    /// server.
-    connect_timeout: Option<Duration>,
+    /// The address to create connections to.
+    address: StreamAddress,
 
-    /// The credential to use for authenticating connections in this pool.
-    credential: Option<Credential>,
+    /// The options used to create connections.
+    connection_options: Option<ConnectionOptions>,
 
     /// Contains the logic for "establishing" a connection. This includes handshaking and
     /// authenticating a connection when it's first created.
     establisher: ConnectionEstablisher,
+}
+
+impl ConnectionManager {
```

Comment:
I assume the implementations here are basically straight copies of those from before, just made async?